strawberryfields/ops.py

```diff
@@ -567,9 +569,16 @@ def apply(self, reg, backend, **kwargs):
         Like :func:`Operation.apply`, but also stores the measurement result in the RegRefs.
         """
         values = super().apply(reg, backend, **kwargs)
-        # measurement can act on multiple modes
+        # convert the returned values into an iterable with the measured modes indexed along
+        # the first axis and shots along second axis (if larger than 1), so that we can assign
+        # register values
+        shots = kwargs.get("shots", 1)
         if self.ns == 1:
-            values = [values]
+            values = [values]  # values is either a scalar, or has shape (shots,)
+        else:
+            if shots > 1:
+                values = values.T  # shape of values would be (shots, num_meas,)
```

Comment:
Here, too, I think values should always be a 2d array even if shots=1, and no transposes should be necessary.

Reply:
@nquesada @co9olguy I think we should try to simplify the way the samples are returned from the backend all the way to the user. Right now the way they are stored changes after almost every function call.

The main question is how should the results be returned to the user.
I don't think returning `None` s for the non-measured systems is very useful.

A few options:

1. A dictionary, mapping indices of measured subsystems to sample lists or 1d-arrays: `samples  = {0: [0.32, 1.43, -0.32], 2: [0, 4, 5]}`
2. An array with just the measured subsystems, leaving it to the user to remember what they are: `samples = array([[0.32, 1.43, -0.32], [0, 4, 5]], dtype=float)`
3. An array like in 2., but with an additional list/1d-array of subsystem indices to label the rows: `modes = [0, 2]`
4. Is there a way in Numpy to label array rows/columns? If there is, 2. and 3. could be combined into a single labeled array.

Reply:
Possibly unpopular opinion: I think the format should be the same even if `shots=1`.

Reply:
I quite like option 2

Reply:
I think it's a bit error-prone, it'd be good if Result also stored the mode numbers the samples correspond to.

Reply:
Like if the user first measures mode 1 and then 0, and assumes the results are returned in temporal order (when in actuality they are returned in subsystem index order).

Reply:
I'm coming around to being more explicit in the `Results` object. And if we use a dictionary, we can still index by mode number (`result[0]` etc.). We just won't be able to more fancy indexing, but maybe we could also provide a separate method to return a numpy array with proper ordering?

Reply:
Agree - of all the options listed, I think (1) and (2) are best (and happy to go with a dictionary). In terms of converting it to a NumPy array, is it as simple as doing `np.array(list(results.values()))`, or will we have to sort it via the keys first? The issue might be Python 3.5, which randomizes returned dictionary values, as opposed to Python 3.6 which returns them in insertion order.

Reply:
To be addressed after this PR is merged. See #107